### DISCUSSION TITLE
Accessibility to blind users

### DISCUSSION THREAD

**Discussion by TPennick** [jamulussoftware/jamulus#1466] [UI Related]
Hi All,
I'm wondering if anyone is working on incorporating accessibility features and fixes into the main Jamulus builds.  I'm aware of some great work by @chigkim and am using his accessibility build for most of my Jamulus interaction.  However, I'm finding that some features of Jamulus work better for me on the main build, e.g. I can read the list of current participants on a server more easily on the main build than on the accessibility build.
My set-up is Windows10 with JAWS for Windows screen reading software, braille display and speech synth using a Behringer UMC 404HD audio interface.  I have a private server on a Ubuntu Linode instance.  Client and server are running Jamulus 3.7.0, and I have both the main, and accessibility clients installed on my PC.
I would like to assist with contributions to the code, particularly to the accessibility, and have software engineering skill, though this is very rusty, and mostly consisted of software maintenance and enhancement on VAX machines back in the 1990s, (though with more recent Linux experience including some use of Git repositories).  Still it would be good to dust these skills off, though would be a major learning curve.
I started to look through the Main.cpp file to get an idea of the structure, and noticed references to accessibility provision which needed to be installed by the installer.  This leads me to wonder whether there is actually accessibility functionality available which I am unaware of.
A basic summary of the problems I am encountering on both main and accessibility builds include:
1. I can't read the various VU and buffering indicators on the main screen.
2. I can't access the faders for the individuals participating in a session.
3. Labelling for the other faders is confusing and not associated with the referenced slider.
4. In the main build, I can't access the table of information about the various servers available on the public list, though can access the name of each server, and the list of current participants.
5. on the main build, I would like to include @chigkim 's alert triggered by a new person joining the session.
I'd be interested to hear from anyone interested in the accessibility of the software, which has obviously become of significant interest to many blind musicians, particularly during the pandemic, both on the subject of current accessibility effort, and also on what's involved in the first few steps towards contributing to the development of Jamulus.  (Would it maybe be possible to have an accessibility category in the Discussion forum?)
Regards and thanks for any thoughts,
Tim Pennick

**Comment by chigkim** [comment]
HI Tim,

1. I can't read the various VU and buffering indicators on the main screen.
The meter is not accessible, but if you go to settings under the misc 
section, you can see overall delay as well as ping time.
Also you can navigate to delay indicator on main window with VO cursor. 
It should say red green etc.
Unfortunately QT made those widget not focussable by keyboard (because I 
guess it doesn't require any user interaction?). However, if you might 
be able to navigate to them by object navigation with nvda, and possibly 
touch cursor with Jaws?
2. I can't access the faders for the individuals participating in a session.

This should work on my accessible build. Just tap to the volume/pan, and 
change them with arrows.

3. Labelling for the other faders is confusing and not associated with 
the referenced slider.
This should be also easy in my accessible build. I made mute/solo to 
include user name, so it should be easy to locate what belongs to who.

I didn't put name in sliders, because it will read out the names 
whenever you change the value. Just mute/solo as anchor for each user, 
and you should be able to locate each user easily. Also you should be 
able to separate title and value using value interface, but I believe it 
requires subclassing the widget which is more complicated, so it's just 
an easy hack around.

4. In the main build, I can't access the table of information about the 
various servers available on the public list, though can access the name 
of each server, and the list of current participants.

The accessible build lets you see name, number of users, ping time, and 
location. However, the down side is that you can't see the users inside 
the table.

The QTreeWidget is broken in Mac with VoiceOver, and you can't focus 
QTreeWidgetItem with VoiceOver. I'm not sure how to fix it, so basically 
I just made each item as a button, so you can at least focus and read 
the value. Unfortunately this breaks the visual of UI, and the main 
reason why I decided make into fork instead of pull request.

Hope that helps, and feel free to make pull request to my fork if you 
have any improvement.


Chi

**Comment by TPennick** [comment]
Hi Chi,

 

Many thanks for your very helpful response.  I’ve been trying to understand why I was struggling to get the results you list below.  I now realise that I was misunderstanding the entries in the personal mix list of current users.  I can now confirm that I can indeed access the pan and fader levels for individuals using your accessibility build.

 

Thanks again for your help,

 

Tim

**Reply by npostavs** [reply]
>  The QTreeWidget is broken in Mac with VoiceOver, and you can't focus QTreeWidgetItem with VoiceOver. I'm not sure how to fix it, so basically I just made each item as a button, so you can at least focus and read the value. Unfortunately this breaks the visual of UI, and the main reason why I decided make into fork instead of pull request.

Is it worth merging in your other changes though?

**Comment by hoffie** [comment]
Hi,

thanks for bringing this up!

I see that @chigkim has already responded to most (all?) of the specific questions, I'll just add some additional comments.

> I'm wondering if anyone is working on incorporating accessibility features and fixes into the main Jamulus builds.

I'm not aware of that, but I would certainly welcome efforts to make the main build more accessible!
Personally, I have a hard time imaging what works good and what doesn't. So what really helps are specific pointers to "this control is confusing to me" or even better "it should rather be like <other working control>".

> I would like to assist with contributions to the code, particularly to the accessibility, and have software engineering skill, though this is very rusty, and mostly consisted of software maintenance and enhancement on VAX machines back in the 1990s, (though with more recent Linux experience including some use of Git repositories). Still it would be good to dust these skills off, though would be a major learning curve.

That sounds good, contributions are certainly welcome! Before you start working on code, it might make sense to write up what the specific problem is and how you're going to solve it. That way, others can chime in and provide guidance if that would be an acceptable PR.

> I started to look through the Main.cpp file to get an idea of the structure, and noticed references to accessibility provision which needed to be installed by the installer. This leads me to wonder whether there is actually accessibility functionality available which I am unaware of.

The fact that many (all?) UI elements have additional accessibility texts (and that @pljones ensures that new controls get them as well ;)) is the only one I can think of. I'm not aware of specific installer modes or check boxes. What exactly do you mean in `main.cpp`?

> A basic summary of the problems I am encountering on both main and accessibility builds include:
> 
>     1. I can't read the various VU and buffering indicators on the main screen.
> 
>     2. I can't access the faders for the individuals participating in a session.
> 
>     3. Labelling for the other faders is confusing and not associated with the referenced slider.
> 
>     4. In the main build, I can't access the table of information about the various servers available on the public list, though can access the name of each server, and the list of current participants.
> 
>     5. on the main build, I would like to include @chigkim 's alert triggered by a new person joining the session.
>        I'd be interested to hear from anyone interested in the accessibility of the software, which has obviously become of significant interest to many blind musicians, particularly during the pandemic, both on the subject of current accessibility effort, and also on what's involved in the first few steps towards contributing to the development of Jamulus. 

As far as I understand, @chigkim has responded to most of that.

> (Would it maybe be possible to have an accessibility category in the Discussion forum?)

In general I'd say yes, although I'd wait a bit more to see of there are many new threads around that topic. As long as it is this single discussion, I'd say that we could keep it here for now?

**Reply by pljones** [reply]
Would it be worth pinning the discussion, so it is, at least, the obvious place to start from?  We can always suggest - as you've done - that specific changes are raised in a separate discussion.

**Comment by chigkim** [comment]
Yes, probably the pan/volume/solo/mute should be ok because it doesn't 
impact how the UI looks.

Should I go ahead and make pull requests?

**Reply by pljones** [reply]
Please do - the usual applies, make sure it's against latest master of the main repo :).

**Comment by TPennick** [comment]
It gets my vote, and can we have at least optionally, the alert when a new participant joins the session.  I find this really valuable.

**Comment by TPennick** [comment]
Not sure how the ‘pinning’ suggested below, works.  Do I have to do something?  Is this like a baseline?

**Reply by pljones** [reply]
The pinning was in response to a post by @hoffie regarding whether to have a specific category in the discussion forum for accessibility.  Pinning is something an administrator can do to a thread to keep it presented prominently above the other discussions.

**Reply by hoffie** [reply]
Good idea (forgot to respond to the initial comment). Have pinned this thread now. Might not be permanent, but may help getting some more attention. :)
